### PR TITLE
Add support for multiple adapters per plugin

### PR DESCRIPTION
**Pull request by dhylands**

### CODE REVIEW

**Comment by dhylands** [comment]
Tested with the zwave adapter (although I only have one dongle).
This also gets rid of all of the test issues that I had by using inproc IPC rather than ipc (which uses named pipes).

File: src/adapters/plugin/adapter-manager-proxy.js [hunk 1]
```diff
@@ -99,6 +102,10 @@ class AdapterManagerProxy extends EventEmitter {
         this.pluginClient.sendNotification(Constants.ADAPTER_UNLOADED, {
           adapterId: adapter.id,
         });
+        return;
+
+      case Constants.UNLOAD_PLUGIN:
```
Comment (hobinjk) [inline src/adapters/plugin/adapter-manager-proxy.js]: Can you add this message to the Adapter IPC documentation in the wiki?

File: src/adapters/plugin/adapter-proxy.js [hunk 2]
```diff
@@ -66,98 +55,26 @@ class AdapterProxy extends Adapter {
...
+   *
+   * @returns a promise which resolves when the adapter has
+   *          finished unloading.
+   */
+  unload() {
+    if (this.deferredUnload) {
+      console.error('AdapterProxy: unload already in progress');
+      return Promise.reject();
+    }
+    this.deferredUnload = new Deferred;
```
Comment (hobinjk) [inline src/adapters/plugin/adapter-proxy.js]: nit: Should have parentheses even if they're optional

File: src/adapters/plugin/adapter-proxy.js [hunk 3]
```diff
@@ -206,6 +123,10 @@ class AdapterProxy extends Adapter {
       return Promise.reject(err);
     }
     this.deferredMock = new Deferred();
+
+    // We need the actual device object when we recolve the promise
```
Comment (hobinjk) [inline src/adapters/plugin/adapter-proxy.js]: nit: resolve

File: src/app-instance.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+/*
+ * app-timestamp.js
```
Comment (hobinjk) [inline src/app-instance.js]: Should this file be named app-timestamp.js then?

File: src/app-instance.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// the purpose of this module.
+
+'use strict';
+
+const process = require('process');
+
+class AppInstance {
+
+  constructor() {
+    this.timestamp = this.mark();
```
Comment (hobinjk) [inline src/app-instance.js]: mark() doesn't return anything, so timestamp will be initialized to undefined. If this is fixed, the call to `mark()` in app.js can be removed

File: src/adapters/plugin/adapter-manager-proxy.js (hunk 1 above)
Comment (dhylands) [reply src/adapters/plugin/adapter-manager-proxy.js]: Yep - was planning on doing that.

File: src/app-instance.js (hunk 4 above)
Comment (dhylands) [reply src/app-instance.js]: I had originally called it app-timestamp and decided to call it app-instance. The fact that it uses a timestamp is really just an implementation detail. I'll fix the comment.

File: src/app-instance.js (hunk 5 above)
Comment (dhylands) [reply src/app-instance.js]: Doh

**Pull request merged**